Improve BenchmarkPartitionedOutputOperator

lukasz-stec opened the pull request:
Add benchmarks for common block types
Add utils to create random pages and blocks
Use OperatorFactories in BenchmarkPartitionedOutputOperator

findepi commented on presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java:
```diff
@@ -75,21 +75,50 @@ public void testJoin()
-                // real count is 25, estimation cannot know all rows are duplicate. TODO create better test for UNION
-                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(0., 1.)));
+                // real count is 25, estimation cannot know all rows are duplicate.
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(1, 1)));
```
i would just remove this test case, what do you think?

findepi commented on presto-tests/src/test/java/com/facebook/presto/tests/TestTpchDistributedStats.java:
```diff
@@ -75,21 +75,50 @@ public void testJoin()
+
+        statisticsAssertion.check("SELECT * FROM orders WHERE o_custkey < 900 INTERSECT SELECT * FROM orders WHERE o_custkey > 600",
+                // TODO fix INTERSECT stats calculation as custkey values distirbution is pretty linear
+                checks -> checks.estimate(OUTPUT_ROW_COUNT, relativeError(4, 5)));
```
That's a lot!

Also, Maybe it would be worthy to add the tests to test-local-stats too so that we know how this runs without exchanges -- maybe the transformations resulting in query plan that is hard to estimate are run _after_ reorder joins would run?

kokosing commented:
I checked tests marked with TODO and they are estimated ok. Currently, with a single range it is difficult to do anything better.

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/PlanStatsAndCostsPrinter.java:
```diff
@@ -0,0 +1,143 @@
+        static final class Context
+        {
+            private final int indent;
+            private final StringBuilder output;
```
`output` is always the same, so it's more an attribute of the visitor than a visiting context.

findepi commented:
I cherry-picked `Tests for stats estimation for UNION, EXCEPT and INTERSECT` (as a fixup) into the CBO branch. I don't know what to do with the other commits.
Maybe for now, keep them on a branch so that they can be easily applied to any code version, as needed for development?

kokosing commented:
Extracted https://github.com/starburstdata/presto/pull/26

lukasz-stec committed:
Refactor BenchmarkPartitionedOutputOperator

Add benchmarks for common block types
Add utils to create random pages and blocks
Use OperatorFactories in BenchmarkPartitionedOutputOperator

lukasz-stec committed:
Use static import for Collections.unmodifiableList

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
         return new RunLengthEncodedBlock(blockBuilder.build(), positionCount);
     }
+
+    public enum Encoding
```
move to the top of class

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+
+    public static DictionaryBlock createRandomDictionaryBlock(Block dictionary, int positionCount, boolean isView)
+    {
+        checkArgument(dictionary.getPositionCount() > 0, format("dictionary's positionCount %d is less than or equal to 0", dictionary.getPositionCount()));
```
`dictionary's positionCount` -> `dictionary position count is less ...`

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+            idsOffset = min(RANDOM.nextInt(dictionary.getPositionCount()), 1);
+        }
+
+        int[] ids = IntStream.range(0, positionCount + idsOffset).map(i -> RANDOM.nextInt(max(dictionary.getPositionCount() / 10, 1))).toArray();
```
move:
```
 .map(
 .toArray(
```
to newlines

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+
+        int idsOffset = 0;
+        if (isView) {
+            idsOffset = min(RANDOM.nextInt(dictionary.getPositionCount()), 1);
```
Why:
```
min(RANDOM.nextInt(dictionary.getPositionCount()), 1)
```
and not some static value, e.g: 10?

In fact, I think we can remove `boolean isView` param as it doesn't really change much from perf POV

lukasz-stec replied:
`IsView` dropped. min here doesn't make sense.

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+            idsOffset = min(RANDOM.nextInt(dictionary.getPositionCount()), 1);
+        }
+
+        int[] ids = IntStream.range(0, positionCount + idsOffset).map(i -> RANDOM.nextInt(max(dictionary.getPositionCount() / 10, 1))).toArray();
```
Why:
```
max(dictionary.getPositionCount() / 10, 1)
```
?
You can just choose random value between `[0, dictionary.getPositionCount())` which will be simpler and more straight forward

lukasz-stec replied:
`max(dictionary.getPositionCount() / 10, 1)` chooses poistions from the first 10% of dictionary.
This doessn't makes sense. changed to 

> random value between `[0, dictionary.getPositionCount())`

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+            int positionCount,
+            float primitiveNullRate,
+            float nestedNullRate,
+            boolean createView,
```
IMO we can remove this param for simplicity

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+                offsets[position + 1] = offsets[position];
+            }
+            else {
+                offsets[position + 1] = offsets[position] + (type instanceof RowType ? 1 : RANDOM.nextInt(ENTRY_SIZE) + 1);
```
could you add short comment what is happening here for a reader that doesn't know how nested blocks work?

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+        return wrappedBlock;
+    }
+
+    static <T> Iterable<T> generateListWithNulls(int positionCount, float nullRate, Supplier<T> valueSupplier)
```
Add `@VisibleForTesting`

sopel39 replied:
make it return `List<T>`

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+
+    public static Block createRandomLongDecimalsBlock(int positionCount, float nullRate)
+    {
+        return createLongDecimalsBlock(generateListWithNulls(positionCount, nullRate,
```
each arg in newline

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+                generateListWithNulls(positionCount, nullRate, () -> generateRandomStringWithLength(maxStringLength)));
+    }
+
+    private static String generateRandomStringWithLength(int length)
```
move it below `generateListWithNulls`

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+    public static RunLengthEncodedBlock createRleBlockWithRandomValue(Block block, int positionCount)
+    {
+        checkArgument(block.getPositionCount() > 0, format("block positions %d is less than or equal to 0", block.getPositionCount()));
+        return new RunLengthEncodedBlock(block.getRegion(block.getPositionCount() / 2, 1), positionCount);
```
you just need one value from underlying block

lukasz-stec replied:
I used `getSingleValueBlock` instead.

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+    {
+        int nullCount = (int) (positionCount * nullRate);
+        if (nullCount == 0 && nullRate > 0) {
+            throw new IllegalArgumentException(format("position count %d too log to have at least one null with rate %f", positionCount, nullRate));
```
`position count %d too log to have at least` the sentence doesn't make sense

lukasz-stec replied:
changed to `position count %d too small to have at least one null with rate %f`

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+    {
+        List<Integer> allNumbers = IntStream.range(0, bound).boxed().collect(toList());
+        Collections.shuffle(allNumbers, RANDOM);
+        return allNumbers.stream().limit(count).collect(toUnmodifiableSet());
```
We use Guava `ImmutableSets` across code so that `ImmutableSet.copyOf` is optimized

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+        // Build isNull and offsets of size positionCount
+        boolean[] isNull = null;
+        Set<Integer> nullPositions = null;
+        if (nestedNullRate > 0) {
```
`chooseNullPositions should work with `nestedNullRate` so `if` here is not needed

lukasz-stec replied:
that's correct with regard to `nullPositions` but this if also creates `isNull` array if needed and this arrays is used later on to create the block.

sopel39 commented on core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:
```diff
@@ -596,4 +626,237 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
+        for (int i = 0; i < positionCount; i++) {
+            result.add(nullPositions.contains(i) ? null : valueSupplier.get());
+        }
+        return unmodifiableList(result);
```
that should return `ImmutableList.copyOf` as we use Guava throughout the code for immutable collections. This way `copyOf` are noops in most cases

lukasz-stec replied:
ImmutableList does not support nulls

lukasz-stec reviewed (commented):
review commentes addressed + rebased on master
